# Bloc 14 — Production-grade & Finalisation
## Reproductibilité, tests, dépendances et verrouillage du système

### Partie A — Mise en pratique

## Étape 1 — Audit environnement et reproductibilité

Cette étape vérifie que l'environnement utilisé pour le projet est
identifiable et reproductible.

Avant le verrouillage final du modèle, il est nécessaire de contrôler :

- la version de Python ;
- les principales dépendances ;
- l'environnement utilisé par les notebooks ;
- la présence des fichiers essentiels ;
- la séparation entre développement et test final.

Aucune donnée de 2026 n'est utilisée.

In [1]:
# ==========================================
# BLOCK 14 — INITIALISATION
# ==========================================

from pathlib import Path

import sys
import os
import platform
import importlib.metadata as metadata

import joblib
import pandas as pd
import numpy as np


PROJECT_DIR = Path(
    r"C:\Users\thiba\OneDrive\Bureau\Projet energie"
)

MODEL_DIR = (
    PROJECT_DIR
    / "models"
)

DATA_DIR = (
    PROJECT_DIR
    / "data_processed"
)

TEST_DIR = (
    PROJECT_DIR
    / "tests"
)


RANDOM_STATE = 42

USE_2026 = False


print(
    "Bloc 14 initialisé."
)

Bloc 14 initialisé.


In [2]:
# ==========================================
# PYTHON ENVIRONMENT
# ==========================================

print(
    "Python executable :",
    sys.executable
)

print(
    "Python version :",
    sys.version
)

print(
    "Platform :",
    platform.platform()
)

Python executable : c:\Users\thiba\AppData\Local\Programs\Python\Python311\python.exe
Python version : 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
Platform : Windows-10-10.0.26200-SP0


In [3]:
assert (
    sys.version_info.major == 3
)

assert (
    sys.version_info.minor == 11
)


print(
    "Python 3.11 : OK"
)

Python 3.11 : OK


In [4]:
# ==========================================
# CORE DEPENDENCIES
# ==========================================

CORE_PACKAGES = [

    "numpy",
    "pandas",
    "scikit-learn",
    "torch",
    "fastapi",
    "uvicorn",
    "streamlit",
    "mlflow",
    "psycopg",
    "fairlearn",
    "pyspark",
]

In [5]:
dependency_rows = []


for package in CORE_PACKAGES:

    try:

        version = metadata.version(
            package
        )

        installed = True

    except metadata.PackageNotFoundError:

        version = None
        installed = False


    dependency_rows.append({

        "package":
            package,

        "installed":
            installed,

        "version":
            version,
    })


dependency_audit = pd.DataFrame(
    dependency_rows
)


display(
    dependency_audit
)

,package,installed,version
0,numpy,True,1.26.4
1,pandas,True,2.3.3
2,scikit-learn,True,1.9.0
3,torch,True,2.12.0
4,fastapi,True,0.141.1
5,uvicorn,True,0.30.0
6,streamlit,True,1.64.0
7,mlflow,True,3.16.1
8,psycopg,True,3.3.6
9,fairlearn,True,0.14.0


In [6]:
missing_packages = (
    dependency_audit[
        ~dependency_audit[
            "installed"
        ]
    ]
)


if len(
    missing_packages
) == 0:

    print(
        "Toutes les dépendances principales sont disponibles."
    )

else:

    display(
        missing_packages
    )

Toutes les dépendances principales sont disponibles.


In [7]:
# ==========================================
# ENVIRONMENT SNAPSHOT
# ==========================================

environment_snapshot = {

    "python_executable":
        sys.executable,

    "python_version":
        platform.python_version(),

    "platform":
        platform.platform(),

    "dependencies":
        dependency_audit,

    "uses_2026":
        False,
}

In [8]:
ENVIRONMENT_PATH = (
    MODEL_DIR
    / "block14_environment.joblib"
)


joblib.dump(
    environment_snapshot,
    ENVIRONMENT_PATH
)


print(
    "Snapshot sauvegardé :",
    ENVIRONMENT_PATH
)

Snapshot sauvegardé : C:\Users\thiba\OneDrive\Bureau\Projet energie\models\block14_environment.joblib


In [9]:
# ==========================================
# ESSENTIAL ARTIFACTS
# ==========================================

essential_artifacts = [

    MODEL_DIR
    / "block7_final_state.joblib",

    MODEL_DIR
    / "block11_final_state.joblib",

    MODEL_DIR
    / "block12_final_state.joblib",

    MODEL_DIR
    / "block13_final_state.joblib",
]

In [10]:
artifact_rows = []


for path in essential_artifacts:

    artifact_rows.append({

        "artifact":
            path.name,

        "exists":
            path.exists(),

        "size_mb":
            (
                path.stat().st_size
                / 1024**2
            )
            if path.exists()
            else np.nan,
    })


artifact_audit = pd.DataFrame(
    artifact_rows
)


display(
    artifact_audit
)

,artifact,exists,size_mb
0,block7_final_state.joblib,True,0.403960
1,block11_final_state.joblib,True,0.008680
2,block12_final_state.joblib,True,0.000361
3,block13_final_state.joblib,True,0.000300


In [11]:
assert (
    artifact_audit[
        "exists"
    ].all()
)


print(
    "Artefacts essentiels : OK"
)

Artefacts essentiels : OK


In [12]:
# ==========================================
# PROJECT STRUCTURE
# ==========================================

required_directories = [

    PROJECT_DIR
    / "api",

    PROJECT_DIR
    / "dashboard",

    PROJECT_DIR
    / "tests",

    PROJECT_DIR
    / "src",

    PROJECT_DIR
    / "models",

    PROJECT_DIR
    / "data_processed",
]

In [13]:
directory_audit = pd.DataFrame({

    "directory": [
        str(
            path.relative_to(
                PROJECT_DIR
            )
        )
        for path in required_directories
    ],

    "exists": [
        path.exists()
        for path in required_directories
    ],
})


display(
    directory_audit
)

,directory,exists
0,api,True
1,dashboard,True
2,tests,True
3,src,True
4,models,True
5,data_processed,True


In [14]:
assert (
    directory_audit[
        "exists"
    ].all()
)


print(
    "Structure projet : OK"
)

Structure projet : OK


In [15]:
# ==========================================
# FINAL TEST PROTECTION
# ==========================================

assert (
    USE_2026
    is False
)


print(
    "2026 toujours protégé : OK"
)

2026 toujours protégé : OK


In [16]:
# ==========================================
# STEP 1 SUMMARY
# ==========================================

step1_summary = pd.DataFrame({

    "Check": [
        "Python 3.11",
        "Core dependencies",
        "Essential artifacts",
        "Project structure",
        "2026 untouched",
    ],

    "Status": [
        sys.version_info.minor == 11,
        missing_packages.empty,
        artifact_audit[
            "exists"
        ].all(),
        directory_audit[
            "exists"
        ].all(),
        USE_2026 is False,
    ],
})


display(
    step1_summary
)

,Check,Status
0,Python 3.11,True
1,Core dependencies,True
2,Essential artifacts,True
3,Project structure,True
4,2026 untouched,True


In [17]:
# ==========================================
# SAVE STEP 1
# ==========================================

STEP1_PATH = (
    MODEL_DIR
    / "block14_step1_audit.joblib"
)


joblib.dump(
    {
        "environment":
            environment_snapshot,

        "artifacts":
            artifact_audit,

        "directories":
            directory_audit,

        "summary":
            step1_summary,

        "uses_2026":
            False,
    },
    STEP1_PATH,
)


print(
    "Étape 1 sauvegardée :",
    STEP1_PATH
)

Étape 1 sauvegardée : C:\Users\thiba\OneDrive\Bureau\Projet energie\models\block14_step1_audit.joblib


### Partie C — Analyse et améliorations guidées par la théorie

La reproductibilité d'un projet de Machine Learning ne dépend pas
uniquement du modèle.

Elle dépend également de l'environnement logiciel utilisé pour produire
les résultats.

On peut représenter une expérience comme :

$$
Résultat
=
f(
Données,
Code,
Paramètres,
Dépendances,
Environnement
).
$$

Une modification d'une dépendance peut donc modifier le comportement du
pipeline même lorsque les données et le code restent identiques.

Le projet conserve par conséquent un snapshot des principales versions
logicielles avant le verrouillage final.

Les artefacts critiques des blocs précédents sont également vérifiés afin
de s'assurer que le pipeline peut être reconstruit sans relancer
inutilement les expériences.

Conclusion :

L'environnement de production du projet est maintenant audité.

Le projet utilise Python 3.11 et les versions des principales dépendances
sont enregistrées.

Les artefacts critiques ainsi que la structure du projet sont contrôlés
avant le verrouillage du modèle.

Aucune donnée de 2026 n'est utilisée à ce stade.

## Étape 2 — Audit final des tests

### Partie A — Mise en pratique

Avant le verrouillage du modèle, la suite de tests du projet est exécutée
afin de vérifier que les différents composants restent cohérents.

Les tests couvrent notamment :

- l'API FastAPI ;
- le dashboard ;
- PostgreSQL ;
- le Bloc 11 ;
- les états finaux des Blocs 12 et 13.

Aucune donnée de 2026 n'est utilisée.

In [18]:
# ==========================================
# CREATE BLOCK 12 / 13 TESTS
# ==========================================

TEST_FILE = (
    TEST_DIR
    / "test_block12_13.py"
)


test_content = r'''
from pathlib import Path

import joblib


PROJECT_DIR = (
    Path(__file__)
    .resolve()
    .parents[1]
)

MODEL_DIR = (
    PROJECT_DIR
    / "models"
)


def test_block12_final_state_exists():

    path = (
        MODEL_DIR
        / "block12_final_state.joblib"
    )

    assert path.exists()


def test_block13_final_state_exists():

    path = (
        MODEL_DIR
        / "block13_final_state.joblib"
    )

    assert path.exists()


def test_2026_is_untouched():

    block12 = joblib.load(
        MODEL_DIR
        / "block12_final_state.joblib"
    )

    block13 = joblib.load(
        MODEL_DIR
        / "block13_final_state.joblib"
    )

    assert (
        block12["uses_2026"]
        is False
    )

    assert (
        block13["uses_2026"]
        is False
    )


def test_final_architecture_decisions():

    block12 = joblib.load(
        MODEL_DIR
        / "block12_final_state.joblib"
    )

    block13 = joblib.load(
        MODEL_DIR
        / "block13_final_state.joblib"
    )

    assert (
        block12["spark_production"]
        is False
    )

    assert (
        block12["postgresql_production"]
        is True
    )

    assert (
        block13["retrained_models"]
        is False
    )

    assert (
        block13["final_rmse"]
        <
        block13["ridge_rmse"]
    )
'''


TEST_FILE.write_text(
    test_content,
    encoding="utf-8",
)


print(
    "Test créé :",
    TEST_FILE
)

Test créé : C:\Users\thiba\OneDrive\Bureau\Projet energie\tests\test_block12_13.py


In [19]:
assert (
    TEST_FILE.exists()
)


print(
    "test_block12_13.py : OK"
)

test_block12_13.py : OK


### Partie C — Analyse et améliorations guidées par la théorie

Les tests automatisés constituent une protection contre les régressions.

Lorsqu'un projet comporte plusieurs composants, une modification locale
peut produire un effet inattendu sur une autre partie du système.

On cherche donc à vérifier des invariants tels que :

$$
uses\_2026 = False,
$$

$$
Spark_{production} = False,
$$

et :

$$
RMSE_{final} < RMSE_{Ridge}.
$$

Les tests permettent ainsi de contrôler automatiquement que certaines
décisions importantes du projet restent respectées.

Les warnings de dépendances sont distingués des erreurs fonctionnelles :
un warning doit être analysé, mais il ne signifie pas nécessairement que
le système est incorrect.

Conclusion :

La suite de tests constitue le dernier contrôle fonctionnel global avant
le verrouillage du système.

Les états finaux des Blocs 12 et 13 sont maintenant également couverts
par des tests simples portant sur les décisions d'architecture et la
protection de l'année 2026.

Le modèle ne sera verrouillé que lorsque la suite complète de tests sera
validée.

### Conclusion — Étape 2

La suite complète de tests du projet a été exécutée avec succès.

Résultat :

- 24 tests réussis sur 24 ;
- aucun test en échec ;
- 3 warnings non bloquants ;
- temps total d'exécution : environ 212.5 secondes.

Les composants suivants ont notamment été validés :

- FastAPI ;
- PostgreSQL ;
- Streamlit ;
- MLflow ;
- Bloc 11 ;
- Bloc 12 ;
- Bloc 13 ;
- protection de l'année 2026.

L'Étape 2 est donc validée avant le verrouillage final du système.

## Étape 3 — Audit des artefacts de production

### Partie A — Mise en pratique

Cette étape identifie les principaux artefacts nécessaires au fonctionnement
et à la reproductibilité du système.

Pour chaque artefact critique, on vérifie :

- son existence ;
- sa taille ;
- sa date de modification ;
- son empreinte SHA-256.

L'empreinte SHA-256 permettra de détecter une modification d'un artefact
après le verrouillage final du projet.

Aucune donnée de 2026 n'est utilisée.

In [20]:
# ==========================================
# CRITICAL PRODUCTION ARTIFACTS
# ==========================================

critical_artifacts = {

    "FastAPI":
        PROJECT_DIR
        / "api"
        / "main.py",

    "Streamlit":
        PROJECT_DIR
        / "dashboard"
        / "app.py",

    "PostgreSQL":
        PROJECT_DIR
        / "src"
        / "database"
        / "postgres.py",

    "Block 7 final state":
        MODEL_DIR
        / "block7_final_state.joblib",

    "Block 11 final state":
        MODEL_DIR
        / "block11_final_state.joblib",

    "Block 12 final state":
        MODEL_DIR
        / "block12_final_state.joblib",

    "Block 13 final state":
        MODEL_DIR
        / "block13_final_state.joblib",
}

In [21]:
# ==========================================
# SHA-256 FUNCTION
# ==========================================

import hashlib


def sha256_file(path):

    hasher = hashlib.sha256()

    with open(
        path,
        "rb",
    ) as file:

        for chunk in iter(
            lambda: file.read(
                1024 * 1024
            ),
            b"",
        ):

            hasher.update(
                chunk
            )

    return hasher.hexdigest()

In [22]:
# ==========================================
# ARTIFACT AUDIT
# ==========================================

from datetime import datetime


artifact_rows = []


for name, path in critical_artifacts.items():

    exists = path.exists()

    artifact_rows.append({

        "artifact":
            name,

        "path":
            str(
                path.relative_to(
                    PROJECT_DIR
                )
            ),

        "exists":
            exists,

        "size_mb":
            (
                path.stat().st_size
                / 1024**2
            )
            if exists
            else np.nan,

        "modified":
            (
                datetime.fromtimestamp(
                    path.stat().st_mtime
                )
            )
            if exists
            else None,

        "sha256":
            (
                sha256_file(
                    path
                )
            )
            if exists
            else None,
    })


production_artifacts = pd.DataFrame(
    artifact_rows
)


display(
    production_artifacts
)

,artifact,path,exists,size_mb,modified,sha256
0,FastAPI,api\main.py,True,0.018218,2026-09-23 20:32:48.327105,83af5eb203cdcc67090ad4fba5f3fa17c866c0b7c7eda8...
1,Streamlit,dashboard\app.py,True,0.034907,2026-09-23 21:30:30.737155,ba327247472830fe0fe3c3dfdbda11bab7c271dcf12732...
2,PostgreSQL,src\database\postgres.py,True,0.001286,2026-09-22 17:23:06.314755,f18c7173ac8b6cc0e7d70f71e39b9ca931b3c1fdc6b381...
3,Block 7 final state,models\block7_final_state.joblib,True,0.403960,2026-09-22 16:19:22.261618,6f3ebb947105cf5760a3284ef05724bed726521a508ed6...
4,Block 11 final state,models\block11_final_state.joblib,True,0.008680,2026-09-24 15:40:27.035996,8d7a569a6fd25c34247511cb0f8f42d66cc59c154c1d17...
5,Block 12 final state,models\block12_final_state.joblib,True,0.000361,2026-09-25 20:17:52.280737,5b21d2d1b50779f38fde2af5f5dfa43972d157bf673a59...
6,Block 13 final state,models\block13_final_state.joblib,True,0.000300,2026-09-25 20:40:02.311242,b8806eb4953a3d8a9a7e6b5a622c8e05564439f976391b...


In [23]:
# ==========================================
# ARTIFACT VALIDATION
# ==========================================

assert (
    production_artifacts[
        "exists"
    ].all()
)


assert (
    production_artifacts[
        "sha256"
    ].notna().all()
)


print(
    "Artefacts critiques : OK"
)

Artefacts critiques : OK


In [24]:
# ==========================================
# CRITICAL FOOTPRINT
# ==========================================

critical_size_mb = float(
    production_artifacts[
        "size_mb"
    ].sum()
)


print(
    "Empreinte des artefacts critiques :",
    f"{critical_size_mb:.3f} MB"
)

Empreinte des artefacts critiques : 0.468 MB


In [28]:
# ==========================================
# MLFLOW CHAMPION AUDIT
# ==========================================

import mlflow

from mlflow.tracking import MlflowClient


TRACKING_URI = (
    "http://127.0.0.1:5000"
)

MLFLOW_MODEL_NAME = (
    "energy-demand-forecasting-model"
)


# Même configuration que tests/test_mlflow.py
mlflow.set_tracking_uri(
    TRACKING_URI
)

mlflow.set_registry_uri(
    TRACKING_URI
)


client = MlflowClient(
    tracking_uri=TRACKING_URI,
    registry_uri=TRACKING_URI,
)


print(
    "Tracking URI :",
    mlflow.get_tracking_uri()
)

print(
    "Registry URI :",
    mlflow.get_registry_uri()
)


champion = (
    client
    .get_model_version_by_alias(
        MLFLOW_MODEL_NAME,
        "champion",
    )
)


print(
    "\nModel name :",
    champion.name
)

print(
    "Version :",
    champion.version
)

print(
    "Run ID :",
    champion.run_id
)

print(
    "Status :",
    champion.status
)

print(
    "RMSE :",
    champion.tags.get(
        "rmse"
    )
)

print(
    "2026 used :",
    champion.tags.get(
        "2026_used"
    )
)

Tracking URI : http://127.0.0.1:5000
Registry URI : http://127.0.0.1:5000

Model name : energy-demand-forecasting-model
Version : 4
Run ID : b388c351a60e4cd9a9d70252b65dce6b
Status : READY
RMSE : 176.94256118746154
2026 used : false


In [31]:
# ==========================================
# CHAMPION METADATA
# ==========================================

champion_metadata = {

    "model_name":
        champion.name,

    "alias":
        "champion",

    "version":
        champion.version,

    "run_id":
        champion.run_id,

    "source":
        champion.source,

    "status":
        champion.status,

    "rmse":
        float(
            champion.tags[
                "rmse"
            ]
        ),

    "uses_2026":
        champion.tags[
            "2026_used"
        ],

    "tracking_uri":
        TRACKING_URI,
}


champion_metadata

{'model_name': 'energy-demand-forecasting-model',
 'alias': 'champion',
 'version': '4',
 'run_id': 'b388c351a60e4cd9a9d70252b65dce6b',
 'source': 'models:/m-9c496614e11d403aa1d8dbb4aebe6399',
 'status': 'READY',
 'rmse': 176.94256118746154,
 'uses_2026': 'false',
 'tracking_uri': 'http://127.0.0.1:5000'}

In [32]:
# ==========================================
# SAVE STEP 3
# ==========================================

STEP3_PATH = (
    MODEL_DIR
    / "block14_step3_artifact_audit.joblib"
)


step3_state = {

    "artifacts":
        production_artifacts,

    "critical_size_mb":
        critical_size_mb,

    "mlflow_champion":
        champion_metadata,

    "test_suite": {

        "passed":
            24,

        "failed":
            0,

        "warnings":
            3,
    },

    "uses_2026":
        False,
}


joblib.dump(
    step3_state,
    STEP3_PATH,
)


print(
    "Étape 3 sauvegardée :",
    STEP3_PATH
)

Étape 3 sauvegardée : C:\Users\thiba\OneDrive\Bureau\Projet energie\models\block14_step3_artifact_audit.joblib


In [33]:
# ==========================================
# STEP 3 VALIDATION
# ==========================================

assert (
    len(
        production_artifacts
    )
    ==
    len(
        critical_artifacts
    )
)


assert (
    production_artifacts[
        "exists"
    ].all()
)


assert (
    production_artifacts[
        "sha256"
    ].notna().all()
)


assert (
    champion_metadata[
        "model_name"
    ]
    ==
    "energy-demand-forecasting-model"
)


assert (
    champion_metadata[
        "alias"
    ]
    ==
    "champion"
)


assert np.isclose(
    champion_metadata[
        "rmse"
    ],
    176.94256118746154,
)


assert (
    champion_metadata[
        "uses_2026"
    ]
    ==
    "false"
)


assert (
    champion_metadata[
        "status"
    ]
    ==
    "READY"
)


assert (
    step3_state[
        "uses_2026"
    ]
    is False
)


print(
    "Étape 3 : validation OK"
)

Étape 3 : validation OK


## Étape 4 — Dépendances et nettoyage des warnings

### Partie A — Mise en pratique

La suite complète contient actuellement 24 tests réussis et trois warnings.

L'objectif de cette étape n'est pas de modifier agressivement
l'environnement, mais de distinguer :

- les warnings corrigeables sans risque ;
- les warnings provenant des dépendances tierces ;
- les warnings acceptés temporairement.

Toute modification de dépendance doit être suivie d'une nouvelle
exécution de la suite de tests.

In [34]:
# ==========================================
# DEPENDENCY SNAPSHOT
# ==========================================

import sys
import importlib.metadata as metadata


packages_to_lock = [
    "numpy",
    "pandas",
    "scikit-learn",
    "torch",
    "fastapi",
    "starlette",
    "httpx",
    "mlflow",
    "pydantic",
    "opentelemetry-api",
    "opentelemetry-sdk",
    "psycopg",
    "streamlit",
    "fairlearn",
    "pyspark",
]


dependency_snapshot = {
    package: (
        metadata.version(package)
        if package
        in {
            dist.metadata["Name"]
            for dist in metadata.distributions()
            if dist.metadata["Name"]
        }
        else None
    )
    for package in packages_to_lock
}


print(
    "Python :",
    sys.version.split()[0]
)

for package, version in dependency_snapshot.items():

    print(
        f"{package:25s}",
        version
    )

Python : 3.11.9
numpy                     1.26.4
pandas                    2.3.3
scikit-learn              1.9.0
torch                     2.12.0
fastapi                   0.141.1
starlette                 1.7.0
httpx                     0.28.1
mlflow                    3.16.1
pydantic                  2.9.0
opentelemetry-api         1.42.1
opentelemetry-sdk         1.42.1
psycopg                   3.3.6
streamlit                 1.64.0
fairlearn                 0.14.0
pyspark                   4.2.0


In [35]:
# ==========================================
# WARNING AUDIT
# ==========================================

warning_audit = pd.DataFrame([
    {
        "warning":
            "Starlette TestClient / httpx",

        "origin":
            "FastAPI / Starlette",

        "action":
            "Install httpx2",

        "status":
            "fixed_or_tested",

        "blocking":
            False,
    },

    {
        "warning":
            "PromptModelConfig.model_name",

        "origin":
            "MLflow / Pydantic",

        "action":
            "No local modification",

        "status":
            "accepted_external",

        "blocking":
            False,
    },

    {
        "warning":
            "SelectableGroups deprecated",

        "origin":
            "OpenTelemetry",

        "action":
            "No dependency upgrade before lock",

        "status":
            "accepted_external",

        "blocking":
            False,
    },
])


display(
    warning_audit
)

,warning,origin,action,status,blocking
0,Starlette TestClient / httpx,FastAPI / Starlette,Install httpx2,fixed_or_tested,False
1,PromptModelConfig.model_name,MLflow / Pydantic,No local modification,accepted_external,False
2,SelectableGroups deprecated,OpenTelemetry,No dependency upgrade before lock,accepted_external,False


In [36]:
# ==========================================
# SAVE STEP 4
# ==========================================

STEP4_PATH = (
    MODEL_DIR
    / "block14_step4_dependencies.joblib"
)


step4_state = {

    "python_version":
        sys.version.split()[0],

    "dependencies":
        dependency_snapshot,

    "warnings":
        warning_audit,

    "critical_warnings":
        0,

    "uses_2026":
        False,
}


joblib.dump(
    step4_state,
    STEP4_PATH,
)


print(
    "Étape 4 sauvegardée :",
    STEP4_PATH
)

Étape 4 sauvegardée : C:\Users\thiba\OneDrive\Bureau\Projet energie\models\block14_step4_dependencies.joblib


In [37]:
# ==========================================
# STEP 4 VALIDATION
# ==========================================

assert (
    sys.version_info.major,
    sys.version_info.minor,
) == (
    3,
    11,
)


assert (
    step4_state[
        "critical_warnings"
    ]
    ==
    0
)


assert (
    warning_audit[
        "blocking"
    ].sum()
    ==
    0
)


assert (
    step4_state[
        "uses_2026"
    ]
    is False
)


print(
    "Étape 4 : validation OK"
)

Étape 4 : validation OK


### Partie C — Analyse et améliorations guidées par la théorie

Un environnement reproductible ne signifie pas nécessairement qu'il
doit produire zéro warning.

Il faut distinguer :

1. une erreur fonctionnelle ;
2. un warning annonçant une incompatibilité future ;
3. un warning interne à une dépendance tierce.

Modifier plusieurs bibliothèques uniquement pour supprimer un warning peut
augmenter le risque de régression.

Dans ce projet, la priorité avant le verrouillage final est donc :

$$
\text{stabilité fonctionnelle}
>
\text{absence esthétique de warnings}.
$$

Les warnings externes non bloquants sont documentés, tandis que les
modifications de dépendances susceptibles d'affecter le pipeline sont
reportées après le verrouillage final.

Conclusion :

L'environnement Python 3.11 est maintenant audité et les avertissements
restants sont classifiés.

Aucun warning restant n'affecte actuellement le fonctionnement du pipeline.
La suite complète avait précédemment validé 24 tests sur 24.

Les dépendances tierces ne sont pas modifiées inutilement avant le
verrouillage final et aucune donnée de 2026 n'a été utilisée.

## Étape 5 — Final Model Lock

### Partie A — Mise en pratique

Cette étape verrouille officiellement la version finale du système avant
toute utilisation de l'année 2026.

Le verrouillage conserve :

- l'identité du modèle final ;
- les performances obtenues sur 2025 ;
- le champion MLflow ;
- les décisions d'architecture ;
- les versions principales des dépendances ;
- les empreintes SHA-256 des artefacts critiques ;
- l'état de la suite de tests.

Après ce verrouillage, aucun choix de modèle ou d'hyperparamètre ne devra
être modifié à partir des résultats obtenus sur 2026.

In [38]:
# ==========================================
# LOAD FINAL STATES
# ==========================================

block7_state = joblib.load(
    MODEL_DIR
    / "block7_final_state.joblib"
)

block11_state = joblib.load(
    MODEL_DIR
    / "block11_final_state.joblib"
)

block12_state = joblib.load(
    MODEL_DIR
    / "block12_final_state.joblib"
)

block13_state = joblib.load(
    MODEL_DIR
    / "block13_final_state.joblib"
)


print(
    "Bloc 7 :",
    block7_state["model_name"]
)

print(
    "RMSE 2025 :",
    block7_state["rmse_2025"]
)

print(
    "Bloc 12 utilise 2026 :",
    block12_state["uses_2026"]
)

print(
    "Bloc 13 utilise 2026 :",
    block13_state["uses_2026"]
)

Bloc 7 : Ridge + MLP1 + Patch Transformer + Refined Event Expert
RMSE 2025 : 176.94256118746154
Bloc 12 utilise 2026 : False
Bloc 13 utilise 2026 : False


In [39]:
# ==========================================
# FINAL PERFORMANCE
# ==========================================

final_metrics = (
    block7_state[
        "metrics_2025"
    ].copy()
)


final_metrics

{'MSE': 31308.66995957857,
 'RMSE': 176.94256118746154,
 'MAE': 127.8516195648178,
 'MAPE_%': 0.25761173413837657,
 'R2': 0.9997223768098071}

In [40]:
EXPECTED_RMSE = (
    176.94256118746154
)


assert np.isclose(
    block7_state[
        "rmse_2025"
    ],
    EXPECTED_RMSE,
)


print(
    "RMSE final verrouillé :",
    block7_state[
        "rmse_2025"
    ]
)

RMSE final verrouillé : 176.94256118746154


In [41]:
# ==========================================
# 2026 PROTECTION
# ==========================================

assert (
    block11_state[
        "uses_2026"
    ]
    is False
)

assert (
    block12_state[
        "uses_2026"
    ]
    is False
)

assert (
    block13_state[
        "uses_2026"
    ]
    is False
)

assert (
    champion_metadata[
        "uses_2026"
    ]
    ==
    "false"
)


print(
    "Protection 2026 : OK"
)

Protection 2026 : OK


In [42]:
# ==========================================
# FINAL LOCK CONFIGURATION
# ==========================================

final_configuration = {

    "project":
        "Probabilistic & Causal Machine Learning "
        "for Energy Demand",

    "lock_version":
        "1.0",

    "model":
        block7_state[
            "model_name"
        ],

    "main_components": [
        "Ridge",
        "MLP",
        "Patch Transformer",
        "Refined Event Expert",
    ],

    "evaluation_period":
        "2025",

    "final_rmse_2025":
        float(
            block7_state[
                "rmse_2025"
            ]
        ),

    "mlflow_model":
        champion_metadata[
            "model_name"
        ],

    "mlflow_alias":
        champion_metadata[
            "alias"
        ],

    "mlflow_version":
        champion_metadata[
            "version"
        ],

    "mlflow_run_id":
        champion_metadata[
            "run_id"
        ],

    "spark_production":
        block12_state[
            "spark_production"
        ],

    "postgresql_production":
        block12_state[
            "postgresql_production"
        ],

    "retrained_block13":
        block13_state[
            "retrained_models"
        ],

    "python_version":
        step4_state[
            "python_version"
        ],

    "tests_passed":
        24,

    "tests_failed":
        0,

    "uses_2026":
        False,
}

In [43]:
# ==========================================
# FINAL ARTIFACT MANIFEST
# ==========================================

final_artifact_manifest = (
    production_artifacts.copy()
)


assert (
    final_artifact_manifest[
        "exists"
    ].all()
)

assert (
    final_artifact_manifest[
        "sha256"
    ].notna().all()
)


display(
    final_artifact_manifest[
        [
            "artifact",
            "path",
            "size_mb",
            "sha256",
        ]
    ]
)

,artifact,path,size_mb,sha256
0,FastAPI,api\main.py,0.018218,83af5eb203cdcc67090ad4fba5f3fa17c866c0b7c7eda8...
1,Streamlit,dashboard\app.py,0.034907,ba327247472830fe0fe3c3dfdbda11bab7c271dcf12732...
2,PostgreSQL,src\database\postgres.py,0.001286,f18c7173ac8b6cc0e7d70f71e39b9ca931b3c1fdc6b381...
3,Block 7 final state,models\block7_final_state.joblib,0.403960,6f3ebb947105cf5760a3284ef05724bed726521a508ed6...
4,Block 11 final state,models\block11_final_state.joblib,0.008680,8d7a569a6fd25c34247511cb0f8f42d66cc59c154c1d17...
5,Block 12 final state,models\block12_final_state.joblib,0.000361,5b21d2d1b50779f38fde2af5f5dfa43972d157bf673a59...
6,Block 13 final state,models\block13_final_state.joblib,0.000300,b8806eb4953a3d8a9a7e6b5a622c8e05564439f976391b...


In [44]:
# ==========================================
# LOCK TIMESTAMP
# ==========================================

from datetime import datetime, timezone


locked_at_utc = (
    datetime.now(
        timezone.utc
    )
    .isoformat()
)


print(
    "Locked at UTC :",
    locked_at_utc
)

Locked at UTC : 2026-09-27T15:43:01.947800+00:00


In [45]:
# ==========================================
# FINAL MODEL LOCK
# ==========================================

final_lock = {

    "locked_at_utc":
        locked_at_utc,

    "configuration":
        final_configuration,

    "metrics_2025":
        final_metrics,

    "mlflow_champion":
        champion_metadata,

    "artifact_manifest":
        final_artifact_manifest,

    "dependencies":
        step4_state[
            "dependencies"
        ],

    "warning_audit":
        step4_state[
            "warnings"
        ],

    "block11_state":
        block11_state,

    "block12_state":
        block12_state,

    "block13_state":
        block13_state,

    "model_selection_frozen":
        True,

    "hyperparameters_frozen":
        True,

    "uses_2026":
        False,
}

In [46]:
# ==========================================
# SAVE FINAL LOCK
# ==========================================

FINAL_LOCK_PATH = (
    MODEL_DIR
    / "block14_final_model_lock.joblib"
)


joblib.dump(
    final_lock,
    FINAL_LOCK_PATH,
)


print(
    "Final Model Lock sauvegardé :",
    FINAL_LOCK_PATH
)

Final Model Lock sauvegardé : C:\Users\thiba\OneDrive\Bureau\Projet energie\models\block14_final_model_lock.joblib


In [47]:
# ==========================================
# FINAL LOCK SHA-256
# ==========================================

final_lock_sha256 = (
    sha256_file(
        FINAL_LOCK_PATH
    )
)


print(
    "Final lock SHA-256 :",
    final_lock_sha256
)

Final lock SHA-256 : 40fb9b451409f5ee86bc520350a9a107abc29b17020c9de0ae8c36dbbf7c169f


In [48]:
FINAL_LOCK_HASH_PATH = (
    MODEL_DIR
    / "block14_final_model_lock.sha256"
)


FINAL_LOCK_HASH_PATH.write_text(
    final_lock_sha256,
    encoding="utf-8",
)


print(
    "Hash sauvegardé :",
    FINAL_LOCK_HASH_PATH
)

Hash sauvegardé : C:\Users\thiba\OneDrive\Bureau\Projet energie\models\block14_final_model_lock.sha256


In [49]:
# ==========================================
# FINAL MODEL LOCK VALIDATION
# ==========================================

assert (
    FINAL_LOCK_PATH.exists()
)

assert (
    FINAL_LOCK_HASH_PATH.exists()
)

assert (
    final_configuration[
        "tests_passed"
    ]
    ==
    24
)

assert (
    final_configuration[
        "tests_failed"
    ]
    ==
    0
)

assert np.isclose(
    final_configuration[
        "final_rmse_2025"
    ],
    176.94256118746154,
)

assert (
    final_configuration[
        "spark_production"
    ]
    is False
)

assert (
    final_configuration[
        "postgresql_production"
    ]
    is True
)

assert (
    final_lock[
        "model_selection_frozen"
    ]
    is True
)

assert (
    final_lock[
        "hyperparameters_frozen"
    ]
    is True
)

assert (
    final_lock[
        "uses_2026"
    ]
    is False
)


print(
    "FINAL MODEL LOCK : OK"
)

FINAL MODEL LOCK : OK


### Partie C — Analyse et améliorations guidées par la théorie

Le test final doit rester statistiquement indépendant des décisions prises
pendant le développement.

Si les résultats obtenus sur le jeu de test sont utilisés pour modifier
l'architecture, les hyperparamètres ou les règles d'ensemble, le jeu de
test devient implicitement un jeu de validation.

On cherche au contraire à conserver :

$$
D_{\mathrm{train}},
D_{\mathrm{validation}}
\perp
D_{\mathrm{test\ final}}
$$

au niveau du processus de sélection du modèle.

Dans ce projet, l'année 2025 a été utilisée pour comparer et sélectionner
les différentes configurations.

L'année 2026 est réservée à une seule évaluation finale après verrouillage.

Le Final Model Lock enregistre donc :

- le modèle sélectionné ;
- ses performances de référence ;
- son identité MLflow ;
- l'environnement logiciel ;
- les décisions d'architecture ;
- les empreintes cryptographiques des artefacts ;
- la preuve que 2026 n'a pas encore été utilisée.

Après ouverture de 2026, les résultats devront être rapportés tels quels.
Ils ne devront pas servir à sélectionner une nouvelle version du modèle
dans cette expérience.

Conclusion :

Le système de prévision est maintenant verrouillé avant l'évaluation finale.

La version retenue combine Ridge, MLP, Patch Transformer et Refined Event
Expert. Elle atteint un RMSE d'environ 176.94 MW sur la période de
développement 2025.

L'API, PostgreSQL, le dashboard, MLflow et les blocs de production ont été
validés par la suite automatisée de tests.

Les principaux artefacts disposent d'une empreinte SHA-256, l'environnement
Python est documenté et l'identité exacte du champion MLflow est conservée.

La sélection du modèle et les hyperparamètres sont désormais figés.

L'année 2026 reste encore totalement inutilisée et peut maintenant servir
de test final indépendant.

## Performance Summary — Bloc 14

### Reproductibilité
- Python : 3.11
- dépendances principales auditées ;
- artefacts critiques vérifiés ;
- empreintes SHA-256 enregistrées.

### Tests
- 24 tests réussis ;
- 0 échec ;
- warnings non bloquants documentés.

### Production
- API : FastAPI ;
- base de données : PostgreSQL ;
- dashboard : Streamlit ;
- registry / tracking : MLflow ;
- Spark conservé comme expérimentation ;
- PostgreSQL/Pandas conservé pour le pipeline principal.

### Modèle verrouillé
- Ridge ;
- MLP ;
- Patch Transformer ;
- Refined Event Expert.

### Performance de référence
- RMSE 2025 : ~176.94 MW ;
- amélioration par rapport au Ridge : ~15.6 %.

### Test final
- 2026 non utilisé pendant le développement ;
- modèle et hyperparamètres figés avant ouverture du test ;
- 2026 devient maintenant le jeu de test final indépendant.